### DISCUSSION TITLE
KMP算法（吴泰维）

### DISCUSSION THREAD

**Discussion by zetawu** [CSGrandeur/s-1problem1day1ac#48] [Algorithm]
KMP算法是用于串的模式匹配的算法。

何为串的模式匹配？请看这道题目。

> 给出两个字符串，主串 `S` 和 模式串 `T`，若 `S` 的区间  [ *l ,r* ]子串与  `T` 完全相同，则称 `T` 在 `S` 中出现了，其出现位置为 *l*。
> 请求出 `T` 在 `S` 中所有出现的位置。

这道题就是串的模式匹配的模板题。

这道题其实并不难想。我们只需要对 `S` 中的每个位置都逐一尝试匹配 `T`，若匹配不成功则从下一个位置开始重新匹配。但是这样做将会出现两层循环，时间复杂度将会达到 `O(n*m)` （n为 `S` 的长度，m为 `T` 的长度）。当数据量过大时，很容易出现超时的问题。

**温馨提示：独立做出此题再阅读此文章，效果更佳。**

而使用KMP算法解决此题，时间复杂度为`O(n+m)` ，在很大程度上避免了超时的风险。

那么KMP算法是怎么实现的？接下来我们要谈的就是这个问题。

 ## KMP算法的实现

先上代码。此代码为本文开头中所给题目的答案（用KMP算法实现）

```c++
#include<bits/stdc++.h>
using namespace std;
char s[1000000], t[1000000];
int i, j, m, n, ansi = 0, next[1000000], ans[10000000];
int main()
{
    memset(next, 0, sizeof(next));
    memset(ans, 0, sizeof(ans));
    scanf("%s%s", s + 1, t + 1);//小技巧：这样输入的字符串可从s[1] t[1]开始存储
    n = strlen(s + 1);//与输入时对应
    m = strlen(t + 1);

    j = 0;
    for(i = 2; i <= m; i++)//求next数组
    {
        while(j != 0 && t[i] != t[j + 1])
            j = next[j];
        if(t[i] == t[j + 1])
            j++;
        next[i] = j;
    }

    j = 0;
    for(i = 1; i <= n; i++)//正式开始串的模式匹配
    {
        while(j != 0 && s[i] != t[j + 1])
            j = next[j];
        if(s[i] == t[j + 1])
        {
            if(j + 1 == m)
            {
                ans[ansi] = i - m + 1;
                ansi++;
                j = next[j + 1];
                continue;
            }
            j++;
        }
    }
    for(i = 0; i < ansi; i++)
        cout << ans[i] <<endl;
    return 0;
}
```

KMP算法的实现过程十分巧妙。

在实现KMP算法的过程中，我们只用到了一重循环。当出现失配（主串字符与模式串字符不匹配）的情况下，我们可以根据失配前匹配成功的部分主串和模式串的匹配信息，结合模式串的某些特征，判断已经匹配成功的主串部分是否仍有可能作为模式串在主串中出现的一段。若可能，我们可以顺着这一段继续比较下去，若不可能，我们也能直接从主串中失配的位置重新与模式串开始比较，从而避免了对主串的重复访问。

我们是通过模式串的哪些特征来判断的？一切的重点就在于代码中的 `next` 数组中。

#### `next` 数组所存储的内容是什么？

`next[i]` 所存储的数据为：“从`t[1]` 到 `t[i]` ，长度为`i` 的这一字符串中，**最大相同前后缀**的长度。”

为了更好地了解`next` 数组，我们需要知道，什么是**前缀**、**后缀**。

> 字符串的**前缀**是指字符串的任意首部。比如字符串 “abbc” 的**前缀**有“a” , “ab” , “abb”。
>
> 同样，字符串的任意尾部是字符串的**后缀**，“abbc”的**后缀**有 “c” , “bc” , “bbc”。

需要注意的一点是，本文中提到的“前缀”与“后缀”，并不包括字符串本身。

而**最大相同前后缀**，指的就是一串字符中，相等的前缀与后缀中最长的一对。

以下是几个**最大相同前后缀**长度的例子。

| 字符串 |    最大相同前后缀长度     |
| :----: | :-----------------------: |
|   AA   |   1（前缀“A”和后缀“A”）   |
|  ABAB  |  2（前缀“AB”和后缀“AB”）  |
| ABABA  | 3（前缀“ABA”和后缀“ABA”） |

并非所有字符串都有相同前后缀。

#### 为什么最大相同前后缀的长度能帮助我们更好地进行串的模式匹配？

当我们进行匹配，出现失配时，模式串中失配位置（代码第26行中的 `j+1` ）前的 `j` 位已经成功与主串中失配位置（代码第26行中的 `i` ）前的 `j` 位匹配成功。

因此，对于模式串中从 `t[0]` 到 `t[j]` 的字符所组成的一个字符串（记作 `u` ）而言，由于 `u` 与失配位置之前的主串是完全匹配的，`u` 的所有后缀均可与主串相互匹配。

同时，`u` 也许存在有相等的前后缀。若字符串 `u` 存在有公共前后缀，那么，`u` 中一定存在一个前缀，与主串相匹配。

当我们知道 `u` 的最大公共前后缀时，我们也就能知道 `u` 中与主串相匹配的最大前缀。我们只需将模式串上的该最大前缀对准主串，就可在防止遗漏结果的情况下，以高效率继续进行串的模式匹配。

在串的模式匹配的过程中，出现失配时，我们无须改动遍历主串时所用的指针（上文代码中的 `i` ），只需移动遍历模式串的指针（上文代码中的 `j` ），极大提高了匹配的效率，尤其是当题设给出的主串有大量可与模式串匹配片段时。

#### 如何求 `next` 数组？

基本思路是：以模式串的前后缀分别作为模式串和主串，进行串的模式匹配，并在失配时记录下最大匹配的串长（ 在上文的代码中，最大匹配的串长可用 `i-m+1 ` 计算出）。

特殊的是，我们可以我们在这个过程中也可使用KMP算法中主串模式匹配的方法（参考代码第24~39行）对模式串的前后缀进行模式匹配，实时更新 `next` 数组，并随时使用，以将求 `next` 数组的复杂度简化。

## KMP算法的优越性

由于不涉及遍历主串的指针（`i`）的回溯，KMP算法可将串的模式匹配的时间复杂度降到 `O(n+m)` 。尤其是当主串中出现大量与模式串部分匹配的片段时，KMP算法可极大地提高串模式匹配的效率。

另一方面，也是由于KMP算法不涉及遍历主串的指针（`i`）的回溯，我们没有了存储主串的必要。若输入格式允许，我们完全可以每输入一个主串字符便匹配一次，并以此将空间复杂度降到 `O(m)` （因为只涉及了模式串的数组操作）。这里的 `m` 为模式串的长度。此时，主串的长度已经不再影响空间复杂度了。

因此，当主串规模过大时，使用KMP算法，每输入一个主串字符便匹配一次，可以显著降低空间复杂度，使我们能做出一些原本做不出的题。

[_参考资料_](https://rainiar.top/archives/38)

**Comment by CSGrandeur** [comment]
知乎的这个回答特别好，可以参考学习：

如何更好地理解和掌握 KMP 算法? - 阮行止的回答 - 知乎
https://www.zhihu.com/question/21923021/answer/1032665486
